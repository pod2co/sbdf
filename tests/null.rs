use std::{fs::File, io::Read};

use insta::assert_ron_snapshot;
use sbdf::{BoolArray, ColumnSlice, Sbdf};

#[test]
fn sample_with_null() {
    // Data table pasted from clipboard CSV containing:
    //
    // ```
    // StringColA,DateColB
    // ,2024/01/01
    // Abc,
    // Def,InvalidDateTest
    // ,
    // PreviousRowWasNull,2025/01/05
    // ```
    // with the first column selected as string, and the second column as date.
    //
    // We want to test the null handling for both columns using the "IsInvalid" property.

    let mut file = File::open("tests/samples/null.sbdf").unwrap();
    let mut in_bytes = Vec::new();
    file.read_to_end(&mut in_bytes).unwrap();
    let sbdf = Sbdf::from_bytes(&in_bytes).unwrap();

    insta::assert_ron_snapshot!(sbdf);

    let out_bytes = sbdf.to_bytes().unwrap();
    assert_eq!(in_bytes, out_bytes);

    let is_invalid_property = |col_idx| -> BoolArray {
        let column: &ColumnSlice = &sbdf.table_slices[0].column_slices[col_idx];
        let is_invalid_property = column.properties.is_invalid.as_ref().unwrap();
        is_invalid_property.decode().unwrap()
    };

    let string_is_invalid = is_invalid_property(0);
    assert_ron_snapshot!(string_is_invalid, @r#"
    BoolArray([
      true,
      false,
      false,
      true,
      false,
    ])
    "#);

    let date_is_invalid = is_invalid_property(1);
    assert_ron_snapshot!(date_is_invalid, @r#"
    BoolArray([
      false,
      true,
      true,
      true,
      false,
    ])
    "#);
}
