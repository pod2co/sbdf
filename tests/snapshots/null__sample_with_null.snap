---
source: tests/null.rs
expression: sbdf
---
Sbdf(
  file_header: FileHeader(
    major_version: 1,
    minor_version: 0,
  ),
  table_metadata: TableMetadata(
    metadata: [],
    columns: [
      ColumnMetadata(
        name: "StringColA",
        ty: String,
        other: [],
      ),
      ColumnMetadata(
        name: "DateColB",
        ty: Date,
        other: [],
      ),
    ],
  ),
  table_slices: [
    TableSlice(
      column_slices: [
        ColumnSlice(
          values: Plain(
            value: StringArray(StringArray([
              "",
              "Abc",
              "Def",
              "",
              "PreviousRowWasNull",
            ])),
          ),
          properties: ColumnProperties(
            is_invalid: Some(EncodedBitArray(
              bit_count: 5,
              bytes: [
                144,
              ],
            )),
            error_code: None,
            has_replaced_value: None,
            other: [],
          ),
        ),
        ColumnSlice(
          values: Plain(
            value: DateArray(DateArray([
              63839664000000,
              49923043200000,
              49923043200000,
              49923043200000,
              63871632000000,
            ])),
          ),
          properties: ColumnProperties(
            is_invalid: Some(EncodedBitArray(
              bit_count: 5,
              bytes: [
                112,
              ],
            )),
            error_code: None,
            has_replaced_value: None,
            other: [],
          ),
        ),
      ],
    ),
  ],
)
